### ISSUE TITLE
Support for packaging types other than war and gwt-app is missing

### ISSUE THREAD

**Issue by fhossfel** [GoogleContainerTools/jib#2562]
UPDATE: At the moment Jib only knows about the packaging types "war" and "gwt-app". Everything else is put in <approot>/lib or <approot>classpath. Since there are other deployable artifacts (like ear, ejb etc.) Jib should be able to decide the correct deployment location relative to appRoot for all these packaging types.

See this [comment](https://github.com/GoogleContainerTools/jib/issues/2562#issuecomment-653591670) for details.

---

**Environment**:

- *Jib version: 2.4.0
- *Build tool:* Maven
- *OS:* Windows 10

**Description of the issue**:

**Expected behavior**:

**Steps to reproduce**:

  1. Have a base image that expects a jar in a certain directory (In this case it is a bundle with Camel routes for Servicemix but I don't think this is releveant)
  2. Run the Jib Maven plugin with the configuration below on a non-war project .
  3. The jar file will *not* end up in /deploy but in /deploy/classpath

**`jib-maven-plugin` Configuration**: 
```xml
			    <plugin>
				    <groupId>com.google.cloud.tools</groupId>
				    <artifactId>jib-maven-plugin</artifactId>
				    <version>2.4.0</version>
				    <configuration>
					    <from>
						    <image>${docker.repo}/common/base:${base.image.version}</image>
					    </from>
					    <to>
						    <image>${docker.repo}/test/${project.artifactId}</image>
						    <tags>
							    <tag>b${env.BUILD_NUMBER}</tag>
						    </tags>
					    </to>
					    <container>
							<appRoot>/deploy</appRoot>
					        <entrypoint>INHERIT</entrypoint>
					    </container>
					    <containerizingMode>packaged</containerizingMode>
				    </configuration>		
			    </plugin>
```

**Comment by chanseokoh** [comment]
Hi @fhossfel,

`appRoot` isn't just a place to put an application JAR file but also some others like `/<appRoot>/libs/` where Jib puts library JARs of your project. If the base image is expecting a JAR at `/deploy/`, I think chances are that putting library JARs into `/deploy/libs/` is also likely not what you want. Does your image work without library JARs? Or, you project does not use any dependencies?

Basically, this doesn't really seem the use case that Jib supports. I'd like to understand more about why the base image should have a JAR placed at `/deploy` and what exactly the content and format of the JAR should be. (For example, is it a runnable JAR you can run with `java -jar`? Do you have a main class (`public static void main()`) in your JAR?)

**Comment by fhossfel** [comment]
Thank you, @chanseokoh, for your quick reply.

The base image we are using is an Apache Servicemix with some common bundles preinstalled. We add an OSGI bundle with our Camel routes to it. So the jar containing the routes has to got into the /deploy folder of the Apache Servicemix. That is basically it.

Jib is really perfect for my purpose because it avoids all the hassle of running kaniko just to build the image containing one extra file. It is just this wrong assumption about the sub folder that cause me headaches. For me it would be perfect if there was a switch - like "-Djib.force.approot=true" - to tell Jib that I do not want it to make any assumption about the sub folder under app root.

**Comment by chanseokoh** [comment]
> We add an OSGI bundle with our Camel routes to it. So the jar containing the routes has to got into the /deploy folder of the Apache Servicemix. That is basically it.

Oh, OSGi, I see. I can certainly think of various workarounds, but before I drop any of them:

Could you share your `pom.xml`? (I am mostly interested in dependencies.) I'd like to understand how your project setup is building an OSGi bundle JAR. I can understand that you just need to place the OGSi bundle, but I think Jib is still unnecessarily copying other dependencies you use into the image (at `/deploy/libs/*.jar`) in addition to your OSGi bundle JAR at `/deploy/classpath/`. Have you looked into the image and checked other JAR files at `/deploy/libs/`?

**Comment by fhossfel** [comment]
I have just checked: There are no unnecessary dependencies. It only copies the main artifact. That's perfect (for my purpose)!

**Comment by chanseokoh** [comment]
> I have just checked: There are no unnecessary dependencies. It only copies the main artifact.

That's surprising. Whenever you declare a dependency in `<dependencies>` in `pom.xml`, that dependency should end up sitting in `/deploy/libs/`. So it may be that you only use the standard JDK API (i.e., only `java.*` packages) and no other libraries. (Or, maybe OGSi dependencies are declared/pulled in in some other way and Maven just doesn't recognize them as normal project dependencies?) Just keep in mind going forward that whenever Maven recognizes some dependency JARs, Jib will copy them into an image.

Anyways, this isn't really the use case Jib is designed for. I think it's not likely that we will update our code to support this. Some options I can think of though:
- [Jib CLI](https://github.com/GoogleContainerTools/jib/tree/master/jib-cli)
   - A command-line tool powered by the same jib-core library that Jib Maven plugin uses. Being a pretty general image building tool, adding new files on top of a base image is trivial.
   - Of course, you'll need to run this as a separate command.
   - _However_, the only problem is that we are currently working on significantly changing the UI. (Jib CLI will continue to provide the same functionality, of course.)
- Write your own [Jib extension](https://github.com/GoogleContainerTools/jib-extensions).
   - With an extension, you have full control over how Jib builds an image.
   - Of course, you have to write some code.
- [`<extraDirectories>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#adding-arbitrary-files-to-the-image) feature
   - You'll need to locally prepare an empty directory (e.g., `<project root>/target/jib-extras`) and copy your JAR into it before you can configure `<extraDirectories>` (because `<extraDirectories>` copies _directory contents_ into an image and is not capable of copying a single file). This can easily be done with some plugins such as [`maven-resources-plugin`](https://maven.apache.org/plugins/maven-resources-plugin/examples/copy-resources.html) or [`copy-rename-maven-plugin`](https://coderplus.github.io/copy-rename-maven-plugin/).
   - _However_, the problem is that you'll end up having both `/deploy/classpath/your.jar` and `/deploy/your.jar`. You will also need to use [Layer-Filter extension](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-maven) to delete `/deploy/classpath/your.jar`.

**Comment by fhossfel** [comment]
The dependency issue is easily explained: Since all our dependencies are either part of Apache Servicemix or are deployed there through the base image the scope for alle dependencies is actually "provided". This works perfectly.

Maybe it would be a better solution not to add an extra option but to put artifacts of the packaging type "bundle" not into classpath but directly into the app root. I would be willing to submit a patch for that.

**Comment by chanseokoh** [comment]
The issue is not as simple as just changing the JAR location. Many users depend on the current app structure in the image and the runtime classpath when writing a script (which we did not intend and we wanted to avoid it). For example, https://stackoverflow.com/a/59361658/1701388 It will be a breaking change.

Therefore, if we were to change the app structure, we should introduce a new containerizingMode. But before we can do so, we should evaluate general use cases thoroughly.

**Comment by chanseokoh** [comment]
And if you find yourself having no trouble writing a patch, then it will be easy for you to write a small Jib extension. Have you considered that?

**Comment by fhossfel** [comment]
The problem is not so much the writing of the Job extension. The problem is to distribute it reliably. I could put it into our Maven repo but I prefer to have a build process that does not rely on any piece of infrastructure which may or may not available.
And, well,  it would only solve my problem. I think other might run into the same issue.

**Comment by chanseokoh** [comment]
Actually if we were to resolve this ourselves, we would write an extension. (We have been resolving certain app framework- and customer-specific issues by writing first-party extensions.) If you contribute to the jib-extensions repo, we can publish and maintain it as a first-party extension.

**Comment by fhossfel** [comment]
I spend some time looking through the code. I think my issue is caused by this piece of code:

    /**
       * Gets whether or not the given project is a war project. This is the case for projects with
       * packaging {@code war} and {@code gwt-app}.
       *
       * @return {@code true} if the project is a war project, {@code false} if not
       */
      @Override
      public boolean isWarProject() {
        String packaging = project.getPackaging();
        return "war".equals(packaging) || "gwt-app".equals(packaging);
      }

This only handles war and gwt-app correctly. However, according to the Maven documentation the even Maven core supports pom, jar, maven-plugin, ejb, war, ear, rar. At least an ear or an ebb (with web services for example) should be deployable to an app server in packaged form just like a war or a GWT app. So I think it would be good to rename that function to "isDeployableProject()" and the add ear, ejb and the rest to it. Ideally  there would be a command line with to force that to true if the packaging type is unknown (like in my example a bundle).

P.S.: I will change the topic of the issue to reflect this issue better.

**Comment by loosebazooka** [comment]
I'm not sure the value added for supporting a legacy format like `ear` is worth it for us.

**Comment by chanseokoh** [comment]
I agree. It's really hard to imagine that Jib will have support for legacy formats like `ear` in the future.

Note anyone can introduce a new Maven packaging type, and Jib certainly cannot and does not intend to support every type someone asks for. Jib didn't even support WAR at the beginning, but we determined that supporting WAR was really worth it as there were huge demands. ([The issue "Support WARs"](https://github.com/GoogleContainerTools/jib/issues/431#issue-334212238) garnered 56 thumbs-up).

**Comment by fhossfel** [comment]
I know most packaging types are "legacy" but they are still used. I think at least the core Maven packaging types should be supported to avoid surprises. You could simply add:

@Override
  public boolean isDeployableArtifact() {
    String packaging = project.getPackaging();
    return "war".equals(packaging) || "gwt-app".equals(packaging) || "ejb-app".equals(packaging) || "ear-app".equals(packaging);
  }

For anythin else the existing logic would apply and the user would have to specify a configuration flag.

Thinking about the issue I do not really understand why the bundle end's up in <approot>/classpath in the first place. If the packaging type is unknown, why not put it directly in appRoot? This seems to be the more reasonable apporache except for *.class (and *.properties maybe) For all other file types - especially if they are nore a ;aven artifact - appRoot seems much more appropriate.

**Comment by chanseokoh** [comment]
First of all, from the technical point of view, I'd like to point out that it is not as simple as changing the condition. The `<containerizingMode>packaged` only works for the `jar` packaging type, and this is actually pretty much intended. Currently, Jib is designed to throw `UnsupportedOperationException` when the user sets `<containerizingMode>packaged` on a non-jar project. And it's unclear if we will eventually support `<containerizingMode>packaged` with war; so far, no one has ever asked for it (which has been a bit of a surprise to me to be honest).

Now, the main reason that we don't natively support `<containerizingMode>packaged` for non-jar projects is because Jib is by design a tool to build an image in an opinionated way. We want Jib to build images in the way that we believe is right and help people avoid creating a bad image; Jib isn't just a tool to put arbitrary files as instructed. As an example with war, Jib unpacks a war artifact and creates 5 layers (.class, resources, third-party deps, snapshot deps, and project deps) out of the exploded war contents to build a highly optimized image. We think it is really bad to put a war artifact or a fat/uber jar without unpacking.

So, even if we decided to add support for a new Maven packaging type, we would first spend a lot of time researching on the artifact format to design the right and opinionated way to containerize it. In any case, technically, changing `isWarProject()` alone doesn't magically make Jib work on new types; our code intentionally validates that `<containerizingMode>packaged` is only used on the `jar` type.

> Thinking about the issue I do not really understand why the bundle end's up in /classpath in the first place.

I agree that it didn't have to be in `/classpath` in the first place, but this is what we decided when we designed it, which works and has been working without any single issue. Technically, I don't think there's a problem to change it. However, as I said, many people depend on the current app structure, and we cannot introduce such a breaking change.

**Comment by fhossfel** [comment]
> Currently, Jib is designed to throw UnsupportedOperationException when the user sets <containerizingMode>packaged on a non-jar project. 

This does not happen. It will happily put in the wrong directory. This would have actually saved me a lot of time.

> I agree that it didn't have to be in /classpath in the first place, but this is what we decided when we designed it, which works and has been working without any single issue.

I will go with the plugin extension for the time being but I still thin this is a bug which should be fixed in the next major release."/classpath" is simply not where people want and/or expect there files to end up.

Shall I create a new issue for that?

**Comment by chanseokoh** [comment]
> > Currently, Jib is designed to throw UnsupportedOperationException when the user sets packaged on a non-jar project.
> 
> This does not happen. It will happily put in the wrong directory. This would have actually saved me a lot of time.

Sorry, I should have been more specific. Jib [throws `UnsupportedOperationException` when the user sets `<containerizingMode>packaged` on a war project](https://github.com/GoogleContainerTools/jib/blob/c867dca50a6a5c6e4e9e4c08ca1ab2cbec32b553/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/PluginConfigurationProcessor.java#L665-L674). And note that `war` and `jar` are only the two types Jib supports, hence my claim "on a non-jar project".

> I will go with the plugin extension for the time being but I still thin this is a bug which should be fixed in the next major release."/classpath" is simply not where people want and/or expect there files to end up.

I am afraid I can't really agree on this perspective. `jar` and `war` are the only packaging types Jib supports, and for these two types, our app structure is what we intended to be. The app root is really an internal details of how we containerize files in an opinionated way and not just where you expect a single file or some kind should end up.

**Comment by chanseokoh** [comment]
> I will go with the plugin extension for the time being

As an idea, I suggest a general file-moving extension where you can relocate any file in the image.

**Comment by fhossfel** [comment]
> I am afraid I can't really agree on this perspective. jar and war are the only packaging types Jib supports, and for these two types, our app structure is what we intended to be. The app root is really an internal details of how we containerize files in an opinionated way and not just where you expect a single file or some kind should end up.

An opinionated approach is fine even though I still think that Jib is a lot less useful then it could be.
But to save lots of peoples tome in my humble opion you should really add:

    @override
    public boolean isDeployableArtifact() {
      if ("war".equals(packaging) ||  "gwt-app".equals(packaging) || "jar".equals(packaging)) {
        return "war".equals(packaging) || "gwt-app".equals(packaging); 
      } else {
       throw new UnsupportedOperationException("jar, war and get-app are the only supported packaging types.");
      }
    }

instead of doing just something with the artifact that doesn't make any sense at all.

**Comment by chanseokoh** [comment]
Note that your packaging type and use case are not Jib is designed to support. You are getting the impression that Jib is putting your artifact in a wrong place, Jib is really designed to work in your case, and it is a bug that Jib is putting an artifact in `/classpath`, but that impression is only because you happened to have all dependencies as "provided" scope. Other people can still shoot themselves in their foot; if someone with the same artifact type had some dependencies in their projects, they would end up in `/appRoot/lib/*`, which I think doesn't make sense either. In this case, we say "we don't support this scenario." Your current usage of Jib is really a hacky way that depends on some internal details that are not designed for your environment. If Jib really supported your use case, it would always do the right thing and in an optimized way no matter how the user has configured their project.

**Comment by chanseokoh** [comment]
So, as you suggested, I agree it does make sense to proactively throw `UnsupportedOperationException` if the packaging types are not what Jib is designed to work for.

**Comment by chanseokoh** [comment]
There's an OSGi bundle extension: https://github.com/GoogleContainerTools/jib-extensions/pull/58/files Thanks for publishing it @fhossfel

Back then, I left this issue open for tracking purposes to add a logic to throw `UnsupportedOperationException` if the packaging type is not one of what Jib supports, but now I feel it's not really necessary. I'll close this issue.
